pageserver/src/walreceiver.rs

```diff
@@ -1,170 +1,567 @@
-use lazy_static::lazy_static;
+use itertools::Itertools;
 use postgres_ffi::waldecoder::*;
 use postgres_protocol::message::backend::ReplicationMessage;
 use postgres_types::PgLsn;
 use std::cell::Cell;
-use std::collections::HashMap;
+use std::collections::{HashMap, HashSet};
+use std::num::NonZeroU32;
+use std::ops::ControlFlow;
 use std::str::FromStr;
-use std::sync::Mutex;
+use std::sync::Arc;
 use std::thread_local;
-use std::time::SystemTime;
-use tokio::pin;
+use std::time::{Duration, SystemTime};
+use tokio::{
+    pin,
+    sync::{mpsc, watch},
+    task::JoinHandle,
+};
+use tokio::{select, time};
 use tokio_postgres::replication::ReplicationStream;
 use tokio_postgres::{Client, NoTls, SimpleQueryMessage, SimpleQueryRow};
 use tokio_stream::StreamExt;
 use tracing::*;
+use url::Url;
+
+use etcd_broker::{SkTimelineInfo, SkTimelineSubscription, SkTimelineSubscriptionKind};
 use utils::{
     lsn::Lsn,
     pq_proto::ZenithFeedback,
-    zid::{ZTenantId, ZTenantTimelineId, ZTimelineId},
+    zid::{ZNodeId, ZTenantId, ZTenantTimelineId, ZTimelineId},
 };
 
-//
-// We keep one WAL Receiver active per timeline.
-//
-struct WalReceiverEntry {
-    wal_producer_connstr: String,
-}
-
-lazy_static! {
-    static ref WAL_RECEIVERS: Mutex<HashMap<(ZTenantId, ZTimelineId), WalReceiverEntry>> =
-        Mutex::new(HashMap::new());
-}
-
 thread_local! {
     // Boolean that is true only for WAL receiver threads
     //
     // This is used in `wait_lsn` to guard against usage that might lead to a deadlock.
     pub(crate) static IS_WAL_RECEIVER: Cell<bool> = Cell::new(false);
 }
 
-fn drop_wal_receiver(tenantid: ZTenantId, timelineid: ZTimelineId) {
-    let mut receivers = WAL_RECEIVERS.lock().unwrap();
-    receivers.remove(&(tenantid, timelineid));
-}
-
-// Launch a new WAL receiver, or tell one that's running about change in connection string
-pub fn launch_wal_receiver(
+pub fn init_wal_receiver_main_thread(
     conf: &'static PageServerConf,
-    tenantid: ZTenantId,
-    timelineid: ZTimelineId,
-    wal_producer_connstr: &str,
-) -> Result<()> {
-    let mut receivers = WAL_RECEIVERS.lock().unwrap();
+    mut timeline_updates_receiver: mpsc::UnboundedReceiver<LocalTimelineUpdate>,
+) -> anyhow::Result<()> {
+    let etcd_endpoints = conf.broker_endpoints.clone();
+    ensure!(
+        !etcd_endpoints.is_empty(),
+        "Cannot start wal receiver: etcd endpoints are empty"
+    );
+    let broker_prefix = &conf.broker_etcd_prefix;
+    info!(
+        "Starting wal receiver main thread, etdc endpoints: {}",
+        etcd_endpoints.iter().map(Url::to_string).join(", ")
+    );
 
-    match receivers.get_mut(&(tenantid, timelineid)) {
-        Some(receiver) => {
-            debug!("wal receiver already running, updating connection string");
-            receiver.wal_producer_connstr = wal_producer_connstr.into();
+    thread_mgr::spawn(
+        ThreadKind::WalReceiver,
+        None,
+        None,
+        "WAL receiver main thread",
```

Comment:
Let's name it differently, e.g. WAL receiver launcher thread (and other funcs like walreceiver_main_thread_loop_step), as it is not really walreceiver.